### PR TITLE
fix declaration ambiguity

### PR DESCRIPTION
**Pull request by soelgary**

### CODE REVIEW

File: src/parser/tiger.grm [hunk 1]
```diff
@@ -49,14 +49,20 @@ open Absyn;
 %keyword WHILE FOR TO BREAK LET IN END FUNCTION VAR TYPE ARRAY IF THEN ELSE
          DO OF NIL
 
-%nonassoc FUNCTION VAR TYPE IN
-
+(* This is used as the type and function declaration precedence
+   so it does not interfere with the tokens type and function themselves *)
+%nonassoc DEC
+%nonassoc TYPE
+%nonassoc FUNCTION
```
Comment (nixpulvis) [inline src/parser/tiger.grm]: Do these all need to be separate.

File: src/parser/tiger.grm [hunk 2]
```diff
@@ -49,14 +49,20 @@ open Absyn;
...
 %nonassoc ID
+%nonassoc LBRACK
```
Comment (nixpulvis) [inline src/parser/tiger.grm:57]: Why is this needed?

File: src/parser/tiger.grm [hunk 3]
```diff
@@ -49,14 +49,20 @@ open Absyn;
...
 
+(* Handles the dangling else problem *)
 %nonassoc THEN
-%nonassoc ELSE OF DO
+%nonassoc ELSE DO
 (* An assignment is a valueless expression *)
 %nonassoc ASSIGN
+%nonassoc OF
```
Comment (nixpulvis) [inline src/parser/tiger.grm:64]: ?

File: src/parser/tiger.grm [hunk 4]
```diff
@@ -49,14 +49,20 @@ open Absyn;
...
+%nonassoc DEC
+%nonassoc TYPE
+%nonassoc FUNCTION
 %nonassoc ID
+%nonassoc LBRACK
 
+(* Handles the dangling else problem *)
 %nonassoc THEN
-%nonassoc ELSE OF DO
+%nonassoc ELSE DO
```
Comment (nixpulvis) [inline src/parser/tiger.grm:61]: do ?

File: src/parser/tiger.grm (hunk 2 above)
Comment (soelgary) [inline src/parser/tiger.grm:57]: The main change(%prec DEC) causes a s/r conflict with 

``` text
var : ID .
var : ID . LBRACK ...
```

File: src/parser/tiger.grm (hunk 3 above)
Comment (soelgary) [inline src/parser/tiger.grm:64]: I believe this was one of the conflicts we solved earler today

File: src/parser/tiger.grm (hunk 4 above)
Comment (soelgary) [inline src/parser/tiger.grm:61]: same here as above

File: src/parser/tiger.grm (hunk 1 above)
Comment (soelgary) [inline src/parser/tiger.grm]: `TYPE` and `FUNCTION` can, but `DEC` must have lower precedence

File: src/parser/tiger.grm (hunk 4 above)
Comment (nixpulvis) [inline src/parser/tiger.grm:61]: I swear we got rid of them

Comment (soelgary) [inline src/parser/tiger.grm:61]: Nah, I believe that was ID and LBRACK, but these changes needed them back

**Comment by nixpulvis** [comment]
I'm going to merge this in. I'll think about the presidencies more tomorrow. We just need to do the write up and call it a day.
